packages/graph-explorer/src/core/StateProvider/localDb.ts

```diff
@@ -0,0 +1,223 @@
+import { z } from "zod";
+import { logger } from "../../utils";
+import { deserializeData, serializeData } from "./serializeData";
+import { fromFileToJson, toJsonFileData } from "../../utils/fileData";
+import saveAs from "file-saver";
+
+/*
+
+# Dev Notes
+
+The backup & restore process is a bit complex. Here is how the process breaks
+down.
+
+## Backup Process
+
+This process will create a JSON file with all the contents of the localForage
+storage, along with some extra metadata.
+
+1. Clean up any left over restored entries from previous failed restores.
+2. Gather all data from localForage storage.
+3. Wrap in backup envelope with some metadata about the backup.
+4. Transform data to a format that is safe to serialize to a file.
+5. Transform data in to binary blob data.
+6. Save to file
+
+## File Load Process
+
+This process will convert the binary data in the file to a JSON structure that
+can be restored. The actual restoration process is separate. This step is used
+as a validation of the backup data in the file.
+
+1. Transform binary blob data to file safe JSON data.
+2. Transform file safe JSON data to normal JSON data.
+3. Parse JSON data as backup data.
+
+## Restore Process
+
+Since we don't have transactions, we take a cautious approach to the restore
+process. We attempt to write all entries from the backup file to the localForage
+storage. If this succeeds, then we remove the old entries and rename the
+restored entries.
+
+If there is a failure during the writing of the restore entries then the
+existing database is left in a working state.
+
+1. Clean up any left over restored entries from previous failed restores.
+2. Write all the backup data entries to localForage with "restored" prefix.
+3. Delete all non-restored entries.
+4. Rename all restored entries by removing prefix.
+
+*/
+
+/** Backs up the localDb to a file */
```

Comment:
nit overall: using JSDoc format to include parameters:

```suggestion
/**
 * Restores a backup to the local database in the browser.
 * 
 * @param {SerializedBackup} backup - The backup data to restore.
 * @param {LocalDb} localDb - The local database interface to restore to.
 * @returns {Promise<void>}
 */
```

Reply:
Fair. I should do that.